finality-aleph/src/sync/substrate/justification.rs

```diff
@@ -0,0 +1,166 @@
+    pub fn verify_bytes(
+        &self,
+        justification: &AlephJustification,
+        bytes: Vec<u8>,
+    ) -> Result<(), SessionVerificationError> {
+        use AlephJustification::*;
+        use SessionVerificationError::*;
+        match justification {
+            CommitteeMultisignature(multisignature) => {
+                match self.authority_verifier.is_complete(&bytes, multisignature) {
+                    true => Ok(()),
+                    false => Err(BadMultisignature),
+                }
+            }
+            EmergencySignature(signature) => match self
+                .emergency_signer
+                .as_ref()
+                .ok_or(NoEmergencySigner)?
+                .verify(&bytes, signature)
+            {
+                true => Ok(()),
+                false => Err(BadEmergencySignature),
+            },
+        }
+    }
+}
+
+// This shouldn't be necessary after we remove the legacy justification sync. Then we can also
+// rewrite the implementation above and make it simpler.
+impl<B: Block> LegacyVerifier<B> for SessionVerifier {
+    fn verify(&self, justification: &AlephJustification, hash: B::Hash) -> bool {
+        match self.verify_bytes(justification, hash.encode()) {
+            Ok(()) => true,
+            Err(e) => {
+                warn!(target: "aleph-justification", "Bad justification for block {:?}: {}", hash, e);
+                false
+            }
+        }
+    }
+}
+
+/// Ways in which a justification can fail verification.
+pub enum VerificationError {
+    UnknownSession,
+    Session(SessionVerificationError),
+}
+
+impl From<SessionVerificationError> for VerificationError {
+    fn from(e: SessionVerificationError) -> Self {
+        VerificationError::Session(e)
+    }
+}
+
+impl Display for VerificationError {
+    fn fmt(&self, f: &mut Formatter<'_>) -> Result<(), FmtError> {
+        use VerificationError::*;
+        match self {
+            UnknownSession => write!(f, "justification from unknown session"),
+            Session(e) => write!(f, "{}", e),
+        }
+    }
+}
+
+/// A verifier working for all sessions, although if the session is too new or ancient it will fail
+/// verification -- this is by design.
+pub struct FullVerifier {
+    sessions: ReadOnlySessionMap,
+    period: SessionPeriod,
+}
+
+impl FullVerifier {
+    fn session_verifier<H: SubstrateHeader<Number = BlockNumber>>(
+        &self,
+        header: &H,
+    ) -> Result<SessionVerifier, VerificationError> {
+        let session_id = session_id_from_num(*header.number(), self.period);
+        self.sessions
+            .get(session_id)
+            .map(|authority_data| authority_data.into())
+            .ok_or(VerificationError::UnknownSession)
+    }
+}
+
+impl<H: SubstrateHeader<Number = BlockNumber>> Verifier<Justification<H>> for FullVerifier {
+    type Error = VerificationError;
+
+    fn verify(&self, justification: Justification<H>) -> Result<Justification<H>, Self::Error> {
```

Comment:
nit: this verify seems to be pretty heavy. On each call it locks on SessionMap, clones authorities and creates a new verifier. Still I'm not sure if there is a better way, to create a full verifier.

I guess we would have to somehow cache already existing verifiers and remove them from cache once they are not needed, which is already done in SessionMap :// maybe @DamianStraszak has an idea for that.

Reply:
I agree this is annoying... I'll look into how hard it would be to rewrite ROSM some more to avoid this.

Reply:
Ok, it seems I'll rewrite much more here after all, I'll open a separate PR for some smaller part of this and then write one for the verifier (although it will take longer). This one will likely get closed.